apps/backend/src/app/modules/session/run/run-processor.class.ts

```diff
@@ -154,159 +191,176 @@ export class RunProcessor {
+        throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
+      }
+    }
+  }
+
+  /** @throws {RunValidationError} */
+  validateReplayHeader() {
+    const { session, replayHeader: header } = this;
+
+    if (header.trackType !== session.trackType)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (header.trackNum !== session.trackNum)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (header.magic !== ReplayFile.REPLAY_MAGIC)
+      throw new RunValidationError(ErrorType.BAD_META);
 
-    // If stagesEndAtStageStarts is true then then timestamps should
-    // contain every checkpoint, since the end zone is either the first cp
-    // of the next segment, or the main track's end zone.
-    // If false, the end zone is the final checkpoint of the current segment,
-    // so the timestamps should contain every checkpoint except the last.
-    // Remember, the /end request sent on hitting the end zone doesn't have a
-    // timestamp.
     if (
-      this.trackType === TrackType.STAGE &&
-      !this.zones.tracks.main.stagesEndAtStageStarts
+      header.mapHash.toUpperCase() !==
+      session.mmap.currentVersion.bspHash.toUpperCase()
     )
-      expectedTimestamps -= 1;
+      throw new RunValidationError(ErrorType.BAD_META);
 
-    if (checkpointsRequired && timestamps.length !== expectedTimestamps)
-      throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
-  }
+    if (header.mapName !== session.mmap.name)
+      throw new RunValidationError(ErrorType.BAD_META);
 
-  processReplayFileHeader() {
-    try {
-      this.replay = {
-        ...this.replayFile.readHeader(),
-        overallStats: null,
-        zoneStats: [],
-        frames: []
-      };
-    } catch {
-      throw new RunValidationError(ErrorType.BAD_REPLAY_FILE);
+    if (header.playerSteamID !== this.user.steamID)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (header.gamemode !== session.gamemode)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (!approxEq(header.tickInterval, Tickrates.get(session.gamemode)))
+      throw new RunValidationError(ErrorType.OUT_OF_SYNC);
+
+    const {
+      AllowedSubmitDelayBase,
+      AllowedSubmitDelayIncrement,
+      AllowedSubmitDelayMax,
+      AllowedTimestampDelay
+    } = RunProcessor.Constants;
+
+    // Check timestamps match up with replay start and end times
+    //
+    //   v sessionStart
+    // v runStart                              v header.timestamp    v now
+    // |---------------------------------------|---------------------|
+    // <--------------------------------------><--------------------->
+    //                    ^ headerRunTime                ^ submitDelay
+
+    // Duration of the run (double, in seconds!)
+    const headerRunTime = header.runTime * 1000;
+
+    // When the run was started according to replay file
+    const runStart = header.timestamp - headerRunTime;
```

Comment:
It's not an issue yet but I wanted to note something interesting to keep in mind: some of our movement code improvements involve stopping a movement tick short of completing the full tick, and in the future (probably coinciding with sub-tick timing in general), the timer will reflect these shortened ticks by ticking less than a full tick of time. This means that the end of run true timestamp minus the start of run true timestamp will likely be *slightly* more than the run duration. For most runs this will likely never matter, though it could amount to maybe even a few minutes difference for a 10 hour run. If needed, we could add to the replay data either a record of the tick adjustments or the actual run start timestamp.